File: tests/golden/aoc/2025/10/part2/test_solver.py
from __future__ import annotations

from dataclasses import dataclass
import sys


@dataclass(frozen=True)
class Rat:
    num: int
    den: int

    def sub(self, other: Rat) -> Rat:
        return make_rat(self.num * other.den - other.num * self.den, self.den * other.den)

    def mul(self, other: Rat) -> Rat:
        return make_rat(self.num * other.num, self.den * other.den)

    def scale(self, factor: int) -> Rat:
        return make_rat(self.num * factor, self.den)

    def div(self, other: Rat) -> Rat:
        if other.num == 0:
            raise RuntimeError("Division by zero rational.")
        return make_rat(self.num * other.den, self.den * other.num)

    def is_zero(self) -> bool:
        return self.num == 0

    def is_negative(self) -> bool:
        return self.num < 0

    def is_integer(self) -> bool:
        return self.den == 1

    def as_i64(self) -> int:
        if self.den != 1:
            raise RuntimeError("Expected integer rational.")
        return self.num


@dataclass
class Machine:
    targets: list[int]
    buttons: list[list[int]]


def abs_i64(value: int) -> int:
    if value < 0:
        return -value
    return value


def gcd_i64(left: int, right: int) -> int:
    a = abs_i64(left)
    b = abs_i64(right)
    while b != 0:
        next_value = a % b
        a = b
        b = next_value
    if a == 0:
        return 1
    return a


def make_rat(num: int, den: int) -> Rat:
    if den == 0:
        raise RuntimeError("Zero denominator.")

    reduced_num = num
    reduced_den = den
    if reduced_den < 0:
        reduced_num = -reduced_num
        reduced_den = -reduced_den
    if reduced_num == 0:
        return Rat(0, 1)

    gcd = gcd_i64(reduced_num, reduced_den)
    return Rat(reduced_num // gcd, reduced_den // gcd)


def decode_levels(levels_str: str) -> list[int]:
    parts = levels_str[1:-1].split(",")
    levels = [0] * len(parts)
    index = 0
    for part in parts:
        levels[index] = int(part)
        index += 1
    return levels


def decode_button(button_str: str, width: int) -> list[int]:
    button = [0] * width
    for raw_index in button_str[1:-1].split(","):
        button[int(raw_index)] = 1
    return button


def parse_machine(machine: str) -> Machine:
    parts = machine.split(" ")
    targets = decode_levels(parts[-1])
    buttons = [[0] * len(targets) for _ in range(len(parts) - 2)]
    for part_index in range(1, len(parts) - 1):
        buttons[part_index - 1] = decode_button(parts[part_index], len(targets))
    return Machine(targets, buttons)


def compute_upper_bounds(machine: Machine) -> list[int]:
    upper_bounds = [0] * len(machine.buttons)
    for button_index in range(len(machine.buttons)):
        bound = -1
        for target_index in range(len(machine.targets)):
            if machine.buttons[button_index][target_index] != 0:
                candidate = machine.targets[target_index]
                if bound < 0 or candidate < bound:
                    bound = candidate
        if bound < 0:
            bound = 0
        upper_bounds[button_index] = bound
    return upper_bounds


def sort_buttons_by_upper_bound(machine: Machine, upper_bounds: list[int]) -> None:
    for left_index in range(len(machine.buttons)):
        best_index = left_index
        for right_index in range(left_index + 1, len(machine.buttons)):
            if upper_bounds[right_index] > upper_bounds[best_index]:
                best_index = right_index

        if best_index != left_index:
            tmp_bound = upper_bounds[left_index]
            upper_bounds[left_index] = upper_bounds[best_index]
            upper_bounds[best_index] = tmp_bound

            tmp_button = machine.buttons[left_index]
            machine.buttons[left_index] = machine.buttons[best_index]
            machine.buttons[best_index] = tmp_button


def build_augmented_matrix(machine: Machine) -> list[list[Rat]]:
    matrix: list[list[Rat]] = [None] * len(machine.targets)
    for row_index in range(len(machine.targets)):
        row = [Rat(0, 1) for _ in range(len(machine.buttons) + 1)]
        for col_index in range(len(machine.buttons)):
            row[col_index] = make_rat(machine.buttons[col_index][row_index], 1)
        row[len(machine.buttons)] = make_rat(machine.targets[row_index], 1)
        matrix[row_index] = row
    return matrix


def swap_rows(matrix: list[list[Rat]], left_row: int, right_row: int) -> None:
    temp = matrix[left_row]
    matrix[left_row] = matrix[right_row]
    matrix[right_row] = temp


def rref(matrix: list[list[Rat]], num_buttons: int) -> list[int]:
    pivot_cols = [0] * len(matrix)
    pivot_count = 0
    pivot_row = 0

    for col_index in range(num_buttons):
        if pivot_row >= len(matrix):
            break

        found_row = -1
        for search_row in range(pivot_row, len(matrix)):
            if not matrix[search_row][col_index].is_zero():
                found_row = search_row
                break
        if found_row < 0:
            continue

        if found_row != pivot_row:
            swap_rows(matrix, pivot_row, found_row)

        pivot_value = matrix[pivot_row][col_index]
        for normalize_col in range(col_index, num_buttons + 1):
            matrix[pivot_row][normalize_col] = matrix[pivot_row][normalize_col].div(pivot_value)

        for eliminate_row in range(len(matrix)):
            if eliminate_row == pivot_row:
                continue

            factor = matrix[eliminate_row][col_index]
            if factor.is_zero():
                continue

            for update_col in range(col_index, num_buttons + 1):
                matrix[eliminate_row][update_col] = matrix[eliminate_row][update_col].sub(
                    factor.mul(matrix[pivot_row][update_col])
                )

        pivot_cols[pivot_count] = col_index
        pivot_count += 1
        pivot_row += 1

    return pivot_cols[:pivot_count]


def validate_consistent(matrix: list[list[Rat]], num_buttons: int) -> None:
    for row in matrix:
        if not row_has_coeff(row, num_buttons) and not row[num_buttons].is_zero():
            raise RuntimeError("Machine has no integer solution.")


def row_has_coeff(row: list[Rat], num_buttons: int) -> bool:
    for col_index in range(num_buttons):
        if not row[col_index].is_zero():
            return True
    return False


def collect_free_columns(num_buttons: int, pivot_cols: list[int]) -> list[int]:
    is_pivot = [False] * num_buttons
    for pivot_col in pivot_cols:
        is_pivot[pivot_col] = True

    free_cols = [0] * (num_buttons - len(pivot_cols))
    free_index = 0
    for col_index in range(num_buttons):
        if not is_pivot[col_index]:
            free_cols[free_index] = col_index
            free_index += 1
    return free_cols


def evaluate_cost(
    matrix: list[list[Rat]],
    num_buttons: int,
    pivot_cols: list[int],
    free_cols: list[int],
    free_vals: list[int],
    free_cost: int,
) -> int:
    total_cost = free_cost

    for pivot_index in range(len(pivot_cols)):
        value = matrix[pivot_index][num_buttons]
        for free_index in range(len(free_cols)):
            coeff = matrix[pivot_index][free_cols[free_index]]
            if not coeff.is_zero():
                value = value.sub(coeff.scale(free_vals[free_index]))

        if value.is_negative() or not value.is_integer():
            return -1

        total_cost += value.as_i64()

    return total_cost


def search_best_cost(
    matrix: list[list[Rat]],
    num_buttons: int,
    pivot_cols: list[int],
    free_cols: list[int],
    upper_bounds: list[int],
    free_vals: list[int],
    depth: int,
    free_cost: int,
    best_cost: int,
) -> int:
    current_best = best_cost
    if depth == len(free_cols):
        candidate = evaluate_cost(matrix, num_buttons, pivot_cols, free_cols, free_vals, free_cost)
        if candidate >= 0 and (current_best < 0 or candidate < current_best):
            return candidate
        return current_best

    upper_bound = upper_bounds[free_cols[depth]]
    for current in range(upper_bound + 1):
        next_free_cost = free_cost + current
        if current_best >= 0 and next_free_cost >= current_best:
            continue

        free_vals[depth] = current
        current_best = search_best_cost(
            matrix,
            num_buttons,
            pivot_cols,
            free_cols,
            upper_bounds,
            free_vals,
            depth + 1,
            next_free_cost,
            current_best,
        )

    return current_best


def solve_machine(machine_line: str) -> int:
    machine = parse_machine(machine_line)
    upper_bounds = compute_upper_bounds(machine)
    sort_buttons_by_upper_bound(machine, upper_bounds)

    num_buttons = len(machine.buttons)
    matrix = build_augmented_matrix(machine)
    pivot_cols = rref(matrix, num_buttons)
    validate_consistent(matrix, num_buttons)

    free_cols = collect_free_columns(num_buttons, pivot_cols)
    if len(free_cols) > 3:
        raise RuntimeError("Too many free variables.")

    if len(free_cols) == 0:
        return evaluate_cost(matrix, num_buttons, pivot_cols, free_cols, [], 0)

    best_cost = search_best_cost(
        matrix,
        num_buttons,
        pivot_cols,
        free_cols,
        upper_bounds,
        [0] * len(free_cols),
        0,
        0,
        -1,
    )

    if best_cost < 0:
        raise RuntimeError("Machine has no non-negative integer solution.")
    return best_cost


def main() -> int:
    total_presses = 0
    for line in sys.stdin.read().splitlines():
        total_presses += solve_machine(line)

    print("RESULT:", end="")
    print(total_presses)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/golden/aoc/2025/10/part2/test_solver.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public final class test_solver {
    private static final class Rat {
        final long num;
        final long den;

        Rat(long num, long den) {
            this.num = num;
            this.den = den;
        }

        Rat sub(Rat other) {
            return make_rat(this.num * other.den - other.num * this.den, this.den * other.den);
        }

        Rat mul(Rat other) {
            return make_rat(this.num * other.num, this.den * other.den);
        }

        Rat scale(long factor) {
            return make_rat(this.num * factor, this.den);
        }

        Rat div(Rat other) {
            if (other.num == 0L) {
                throw new IllegalStateException("Division by zero rational.");
            }
            return make_rat(this.num * other.den, this.den * other.num);
        }

        boolean isZero() {
            return this.num == 0L;
        }

        boolean isNegative() {
            return this.num < 0L;
        }

        boolean isInteger() {
            return this.den == 1L;
        }

        long as_i64() {
            if (this.den != 1L) {
                throw new IllegalStateException("Expected integer rational.");
            }
            return this.num;
        }
    }

    private static final class Machine {
        final long[] targets;
        final long[][] buttons;

        Machine(long[] targets, long[][] buttons) {
            this.targets = targets;
            this.buttons = buttons;
        }
    }

    private test_solver() {
    }

    private static long abs_i64(long value) {
        if (value < 0L) {
            return -value;
        }
        return value;
    }

    private static long gcd_i64(long left, long right) {
        long a = abs_i64(left);
        long b = abs_i64(right);
        while (b != 0L) {
            long next = a % b;
            a = b;
            b = next;
        }
        if (a == 0L) {
            return 1L;
        }
        return a;
    }

    private static Rat make_rat(long num, long den) {
        if (den == 0L) {
            throw new IllegalStateException("Zero denominator.");
        }

        long reduced_num = num;
        long reduced_den = den;
        if (reduced_den < 0L) {
            reduced_num = -reduced_num;
            reduced_den = -reduced_den;
        }
        if (reduced_num == 0L) {
            return new Rat(0L, 1L);
        }

        long gcd = gcd_i64(reduced_num, reduced_den);
        return new Rat(reduced_num / gcd, reduced_den / gcd);
    }

    private static long[] decode_levels(String levelsStr) {
        String[] parts = levelsStr.substring(1, levelsStr.length() - 1).split(",");
        long[] levels = new long[parts.length];
        for (int index = 0; index < parts.length; index++) {
            levels[index] = Long.parseLong(parts[index]);
        }
        return levels;
    }

    private static long[] decode_button(String buttonStr, int width) {
        long[] button = new long[width];
        String inner = buttonStr.substring(1, buttonStr.length() - 1);
        if (inner.isEmpty()) {
            return button;
        }
        String[] indices = inner.split(",");
        for (String indexStr : indices) {
            button[Integer.parseInt(indexStr)] = 1L;
        }
        return button;
    }

    private static Machine parse_machine(String machine) {
        String[] parts = machine.split(" ");
        long[] targets = decode_levels(parts[parts.length - 1]);
        long[][] buttons = new long[parts.length - 2][];
        for (int partIndex = 1; partIndex < parts.length - 1; partIndex++) {
            buttons[partIndex - 1] = decode_button(parts[partIndex], targets.length);
        }
        return new Machine(targets, buttons);
    }

    private static long[] compute_upper_bounds(Machine machine) {
        long[] upperBounds = new long[machine.buttons.length];
        for (int buttonIndex = 0; buttonIndex < machine.buttons.length; buttonIndex++) {
            long bound = -1L;
            for (int targetIndex = 0; targetIndex < machine.targets.length; targetIndex++) {
                if (machine.buttons[buttonIndex][targetIndex] != 0L) {
                    long candidate = machine.targets[targetIndex];
                    if (bound < 0L || candidate < bound) {
                        bound = candidate;
                    }
                }
            }
            if (bound < 0L) {
                bound = 0L;
            }
            upperBounds[buttonIndex] = bound;
        }
        return upperBounds;
    }

    private static void sort_buttons_by_upper_bound(Machine machine, long[] upperBounds) {
        for (int leftIndex = 0; leftIndex < machine.buttons.length; leftIndex++) {
            int bestIndex = leftIndex;
            for (int rightIndex = leftIndex + 1; rightIndex < machine.buttons.length; rightIndex++) {
                if (upperBounds[rightIndex] > upperBounds[bestIndex]) {
                    bestIndex = rightIndex;
                }
            }

            if (bestIndex != leftIndex) {
                long tmpBound = upperBounds[leftIndex];
                upperBounds[leftIndex] = upperBounds[bestIndex];
                upperBounds[bestIndex] = tmpBound;

                long[] tmpButton = machine.buttons[leftIndex];
                machine.buttons[leftIndex] = machine.buttons[bestIndex];
                machine.buttons[bestIndex] = tmpButton;
            }
        }
    }

    private static Rat[][] build_augmented_matrix(Machine machine) {
        Rat[][] matrix = new Rat[machine.targets.length][];
        for (int rowIndex = 0; rowIndex < machine.targets.length; rowIndex++) {
            Rat[] row = new Rat[machine.buttons.length + 1];
            for (int colIndex = 0; colIndex < machine.buttons.length; colIndex++) {
                row[colIndex] = make_rat(machine.buttons[colIndex][rowIndex], 1L);
            }
            row[machine.buttons.length] = make_rat(machine.targets[rowIndex], 1L);
            matrix[rowIndex] = row;
        }
        return matrix;
    }

    private static void swap_rows(Rat[][] matrix, int leftRow, int rightRow) {
        Rat[] temp = matrix[leftRow];
        matrix[leftRow] = matrix[rightRow];
        matrix[rightRow] = temp;
    }

    private static long[] rref(Rat[][] matrix, int numButtons) {
        long[] pivotCols = new long[matrix.length];
        int pivotCount = 0;
        int pivotRow = 0;

        for (int colIndex = 0; colIndex < numButtons; colIndex++) {
            if (pivotRow >= matrix.length) {
                break;
            }

            int foundRow = -1;
            for (int searchRow = pivotRow; searchRow < matrix.length; searchRow++) {
                if (!matrix[searchRow][colIndex].isZero()) {
                    foundRow = searchRow;
                    break;
                }
            }
            if (foundRow < 0) {
                continue;
            }

            if (foundRow != pivotRow) {
                swap_rows(matrix, pivotRow, foundRow);
            }

            Rat pivotValue = matrix[pivotRow][colIndex];
            for (int normalizeCol = colIndex; normalizeCol < numButtons + 1; normalizeCol++) {
                matrix[pivotRow][normalizeCol] = matrix[pivotRow][normalizeCol].div(pivotValue);
            }

            for (int eliminateRow = 0; eliminateRow < matrix.length; eliminateRow++) {
                if (eliminateRow == pivotRow) {
                    continue;
                }

                Rat factor = matrix[eliminateRow][colIndex];
                if (factor.isZero()) {
                    continue;
                }

                for (int updateCol = colIndex; updateCol < numButtons + 1; updateCol++) {
                    matrix[eliminateRow][updateCol] = matrix[eliminateRow][updateCol].sub(
                        factor.mul(matrix[pivotRow][updateCol])
                    );
                }
            }

            pivotCols[pivotCount] = colIndex;
            pivotCount = pivotCount + 1;
            pivotRow = pivotRow + 1;
        }

        long[] result = new long[pivotCount];
        for (int index = 0; index < pivotCount; index++) {
            result[index] = pivotCols[index];
        }
        return result;
    }

    private static void validate_consistent(Rat[][] matrix, int numButtons) {
        for (Rat[] row : matrix) {
            if (!row_has_coeff(row, numButtons) && !row[numButtons].isZero()) {
                throw new IllegalStateException("Machine has no integer solution.");
            }
        }
    }

    private static boolean row_has_coeff(Rat[] row, int numButtons) {
        for (int colIndex = 0; colIndex < numButtons; colIndex++) {
            if (!row[colIndex].isZero()) {
                return true;
            }
        }
        return false;
    }

    private static long[] collect_free_columns(int numButtons, long[] pivotCols) {
        boolean[] isPivot = new boolean[numButtons];
        for (long pivotCol : pivotCols) {
            isPivot[(int)pivotCol] = true;
        }

        long[] freeCols = new long[numButtons - pivotCols.length];
        int freeIndex = 0;
        for (int colIndex = 0; colIndex < numButtons; colIndex++) {
            if (!isPivot[colIndex]) {
                freeCols[freeIndex] = colIndex;
                freeIndex = freeIndex + 1;
            }
        }
        return freeCols;
    }

    private static long evaluate_cost(
        Rat[][] matrix,
        int numButtons,
        long[] pivotCols,
        long[] freeCols,
        long[] freeVals,
        long freeCost
    ) {
        long totalCost = freeCost;

        for (int pivotIndex = 0; pivotIndex < pivotCols.length; pivotIndex++) {
            Rat value = matrix[pivotIndex][numButtons];
            for (int freeIndex = 0; freeIndex < freeCols.length; freeIndex++) {
                Rat coeff = matrix[pivotIndex][(int)freeCols[freeIndex]];
                if (!coeff.isZero()) {
                    value = value.sub(coeff.scale(freeVals[freeIndex]));
                }
            }

            if (value.isNegative() || !value.isInteger()) {
                return -1L;
            }

            totalCost = totalCost + value.as_i64();
        }

        return totalCost;
    }

    private static long search_best_cost(
        Rat[][] matrix,
        int numButtons,
        long[] pivotCols,
        long[] freeCols,
        long[] upperBounds,
        long[] freeVals,
        int depth,
        long freeCost,
        long bestCost
    ) {
        long currentBest = bestCost;
        if (depth == freeCols.length) {
            long candidate = evaluate_cost(matrix, numButtons, pivotCols, freeCols, freeVals, freeCost);
            if (candidate >= 0L && (currentBest < 0L || candidate < currentBest)) {
                return candidate;
            }
            return currentBest;
        }

        long upperBound = upperBounds[(int)freeCols[depth]];
        for (long current = 0L; current <= upperBound; current++) {
            long nextFreeCost = freeCost + current;
            if (currentBest >= 0L && nextFreeCost >= currentBest) {
                continue;
            }

            freeVals[depth] = current;
            currentBest = search_best_cost(
                matrix,
                numButtons,
                pivotCols,
                freeCols,
                upperBounds,
                freeVals,
                depth + 1,
                nextFreeCost,
                currentBest
            );
        }

        return currentBest;
    }

    private static long solve_machine(String machineLine) {
        Machine machine = parse_machine(machineLine);
        long[] upperBounds = compute_upper_bounds(machine);
        sort_buttons_by_upper_bound(machine, upperBounds);

        int numButtons = machine.buttons.length;
        Rat[][] matrix = build_augmented_matrix(machine);
        long[] pivotCols = rref(matrix, numButtons);
        validate_consistent(matrix, numButtons);

        long[] freeCols = collect_free_columns(numButtons, pivotCols);
        if (freeCols.length > 3) {
            throw new IllegalStateException("Too many free variables.");
        }

        if (freeCols.length == 0) {
            return evaluate_cost(matrix, numButtons, pivotCols, freeCols, new long[0], 0L);
        }

        long bestCost = search_best_cost(
            matrix,
            numButtons,
            pivotCols,
            freeCols,
            upperBounds,
            new long[freeCols.length],
            0,
            0L,
            -1L
        );

        if (bestCost < 0L) {
            throw new IllegalStateException("Machine has no non-negative integer solution.");
        }
        return bestCost;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        List<String> lines = new ArrayList<>();
        while (true) {
            String line = reader.readLine();
            if (line == null) {
                break;
            }
            lines.add(line);
        }

        long totalPresses = 0L;
        for (String line : lines) {
            totalPresses = totalPresses + solve_machine(line);
        }

        System.out.print("RESULT:");
        System.out.println(totalPresses);
    }
}